### ISSUE TITLE
Port twisted.python.runtime to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5897]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5897|
|Type|enhancement|
|Created|2012-08-21 16:00:18Z|
|Branch|https://github.com/twisted/twisted/tree/runtime-python3-5897-2|

twisted.python.runtime should run on Python 3.

Attachments:

* [runtime.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/043/0431e3fb177800b081eef4e70086ba9dd5ca69bb/e422b05ea5a48bef2c66b861c8c3e5ede16881a6.diff) (1544 bytes) - added by meissenPlate on 2012-08-22 03:05:46Z - Patch to make twisted.python.runtime usable with BOTH 2x and 3x.
* [test_runtime.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/043/0431e3fb177800b081eef4e70086ba9dd5ca69bb/1c3bb5efa37de99a0ae3e85de680c88969a8362f.diff) (515 bytes) - added by meissenPlate on 2012-08-22 03:07:45Z - A test that should fail if the unmodified twisted.python.runtime were to be run under 3x. (Since trial itself doesn't run under 3x I can only say that it passes under 2x trial.)
* [twisted_python_runtime_3x_compatibility.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/043/0431e3fb177800b081eef4e70086ba9dd5ca69bb/198839eaa38426afc9e3981f035ed7137d821c14.diff) (2882 bytes) - added by meissenPlate on 2012-08-25 20:24:47Z - 3x compatibility patch for twisted.python.runtime, with changes to that module's tests included.
* [twisted_python_runtime_3x_compatibility_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/043/0431e3fb177800b081eef4e70086ba9dd5ca69bb/cbb24a0207a56045fa3a89247fb9721739bafe71.diff) (2947 bytes) - added by meissenPlate on 2012-08-25 22:20:52Z - 3x compatibility patch for twisted.python.runtime, with changes to that module's tests included.  Ignore the 3rd file.

<details><summary>Searchable metadata</summary>

```
trac-id__5897 5897
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_runtime_python3_5897_2 branches/runtime-python3-5897-2
branch_author__meissenPlate__exarkun meissenPlate, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345564818000000 1345564818000000
changetime__1346250136000000 1346250136000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|meissenPlate commented|
|-|-|

Nothing too complex, in 3x the module "thread" is "_thread" and "_winreg" is "winreg", so we just need to be a little smart when we import them. I added a test that should fail if the OLD runtime.py were run under 3x Python, but since trial itself doesn't run under 3x I can't test that it fails under those conditions. The NEW runtime.py passes all its tests under 2x and can be run and used under both 2x and 3x.

Running 2to3 on the OLD runtime.py corrects the name of the winreg module, but NOT the name of the thread module (because of the way it's being loaded). Running 2to3 on the NEW runtime.py produces no changes.

Incidentally: I noticed this file uses a variable named "type" which is covering up a Python built in. It sort of bothers me but I left it alone because it isn't hurting anything.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to meissenPlate|
|-|-|

Thanks.  There are a few problems here:

  1. Please attach all changes as a single patch file, not a patch file per modified source file.
  1. Using `__import__` to deal with the name changes is gratuitous.  `import x as y` would be perfectly sufficient.
  1. It would be better _not_ to perform the Python version checks in multiple places, and ideally not inside functions that may be called repeatedly.
  1. The tests need to pass on Python 3.  Since trial is not ported yet, please switch the tests to using the standard library unittests module, add a comment at that code site indicating it should be switched back again and reference a newly filed ticket for doing that work.
  1. The new unit test doesn't make any assertions, so it's not very useful.  The best implementation of `isWinNT` that would make the test pass is `pass`.
  1. `twisted.python.compat._PY3` is the preferred way to check whether the Python runtime is a 3.x version or not.
  1. An explicit decision needs to be made about the string type for all strings in this module.  We are probably leaning towards leaving "" literals alone (that is, letting `str` in Python 2.x be `str` in Python 3.x), but the decision still needs to be finalized.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="meissenPlate's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|meissenPlate commented|
|-|-|

Sorry for being so slow, the source control is not very strong with meissenPlate.

Ignore the 3rd file, typing up this message made me realize I still needed to do a couple things.

''1.   Please attach all changes as a single patch file, not a patch file per modified source file.
2.   Using __import__ to deal with the name changes is gratuitous. import x as y would be perfectly sufficient.
6.   twisted.python.compat._PY3 is the preferred way to check whether the Python runtime is a 3.x version or not.''

Done. Sorry about that.

_4.   The tests need to pass on Python 3. Since trial is not ported yet, please switch the tests to using the standard library unittests module, add a comment at that code site indicating it should be switched back again and reference a newly filed ticket for doing that work._

Makes sense. I've changed twisted.python.test.test_runtime to import TestCase from unittest instead of trial, added a TODO, and added ticket [#5919](https://github.com/twisted/twisted/issues/5919) "Use trial to run the tests for twisted.python.runtime" [http://twistedmatrix.com/trac/ticket/5919]

_5.   The new unit test doesn't make any assertions, so it's not very useful. The best implementation of isWinNT that would make the test pass is pass._

When I wrote this test what I was thinking that the value of the test would be that it would have caught one of the two compatibility problems that twisted.python.runtime had, so I wasn't thinking about making it as tight as I could. I took another look at it and was able to tighten it up some: asserting that the return value was either 1 or 0, and always 0 if the system isn't some sort of Windows.

_7.   An explicit decision needs to be made about the string type for all strings in this module. We are probably leaning towards leaving "" literals alone (that is, letting str in Python 2.x be str in Python 3.x), but the decision still needs to be finalized._ 

You're worried about the bytestring/unicode thing? I agree that this is something worth talking about. I think this file's strings are fine how they are?

3.   It would be better not to perform the Python version checks in multiple places, and ideally not inside functions that may be called repeatedly.

I pulled the choice between _thread and thread out of the function into the class. I could do the same with winreg vs _winreg but we would need to go back to __import__ or imp.load() or something. Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [35422]) Branching to 'runtime-python3-5897'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [35423]) Apply twisted_python_runtime_3x_compatibility_2.diff 

refs [#5897](https://github.com/twisted/twisted/issues/5897)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Thanks!

  1. [There is some trailing whitespace in the patch.](http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto5)
  1. The docstring for `test_isWinNT` doesn't need to use the possessive where it does and that docstring should be wrapped at 79 columns.
  1. Python 2.6 stdlib `unittest.TestCase` does not have `assertIn`, so we need to do it manually.
  1. There are still multiple checks for Python 3.  This is a check that can be done at module initialization time, once.
  1. It appears there are no unit tests for `supportsThreads`.
  1. Now that `twisted.python.runtime` has been ported to Python 3, it needs to be added to `admin/_twistedpython3.py` (and its test module)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [35424]) Address review points

refs [#5897](https://github.com/twisted/twisted/issues/5897)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @itamarst|
|-|-|

[Limited build results (just twisted.python.test.test_runtime)](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/runtime-python3-5897).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

[Complete build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/runtime-python3-5897)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

(In [35436]) Branching to 'runtime-python3-5897-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

Looks good; just needs merging forward and adding the `__future__` imports, which I will do.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst commented|
|-|-|

Oh, and add news file. I also pointed at [#5885](https://github.com/twisted/twisted/issues/5885) instead for using trial.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst set status to `closed`|
|-|-|

(In [35440]) Merge runtime-python3-5897-2

Author: meissenPlate, exarkun
Review: exarkun, itamar
Fixes: [#5897](https://github.com/twisted/twisted/issues/5897)

twisted.python.runtime now works on Python 3.
